### PR TITLE
modules/ignition: unify kubelet/docker bootstrap via tectonic-torcx

### PR DESCRIPTION
**Pull request by lucab**
This PR unifies the bootstrapping process of new nodes, taking care of both kubelet (now) and docker (future CL releases, via torcx).
It introduces a new k8s-node-bootstrap.service, replacing and augmenting  kubelet-env.service.
The full bootstrap flow is now documented at `node-bootstrap-flow.md` (see [rendered version](https://github.com/lucab/tectonic-installer/blob/ups/torcx-bootstrapper/Documentation/dev/node-bootstrap-flow.md)).

Summary of the changes:
 * drop kubelet-env.service and kubelet.env
 * introduce k8s-node-bootstrap.service
 * introduce /etc/kubernetes/installer/kubelet.env

/cc @squeed

### CODE REVIEW

File: platforms/vmware/main.tf [hunk 1]
```diff
@@ -65,6 +72,8 @@ module "masters" {
   kubeconfig              = "${module.bootkube.kubeconfig}"
   private_key             = "${var.tectonic_vmware_ssh_private_key_path}"
   image_re                = "${var.tectonic_image_re}"
+
+  ign_installer_kubelet_env_id = "${module.ignition.installer_kubelet_env_id}"
```
Comment (alekssaul) [inline platforms/vmware/main.tf]: I believe this line is supposed to go under `module "masters"` instead of etcd?

Comment (lucab) [inline platforms/vmware/main.tf]: It is already under "masters", github folding is tricking you into believing it's under "etcd" instead ;)

**Comment by cpanato** [comment]
close/open to refresh the labels

**Comment by sudhaponnaganti** [comment]
This is a release blocker

**Comment by lucab** [comment]
This is mostly green (I think Azure failure is a flake in the cloud API) and ready for review.

File: modules/ignition/assets.tf [hunk 2]
```diff
@@ -48,21 +48,23 @@ data "ignition_systemd_unit" "kubelet" {
...
   vars {
-    kube_version_image_url = "${replace(var.container_images["kube_version"],var.image_re,"$1")}"
-    kube_version_image_tag = "${replace(var.container_images["kube_version"],var.image_re,"$2")}"
-    kubelet_image_url      = "${replace(var.container_images["hyperkube"],var.image_re,"$1")}"
-    kubeconfig_fetch_cmd   = "${var.kubeconfig_fetch_cmd != "" ? "ExecStartPre=${var.kubeconfig_fetch_cmd}" : ""}"
+    bootstrap_upgrade_cl     = "${var.bootstrap_upgrade_cl}"
+    kubeconfig_fetch_cmd     = "${var.kubeconfig_fetch_cmd != "" ? "ExecStartPre=${var.kubeconfig_fetch_cmd}" : ""}"
+    tectonic_torcx_image_url = "${replace(var.container_images["tectonic_torcx"],var.image_re,"$1")}"
+    tectonic_torcx_image_tag = "${replace(var.container_images["tectonic_torcx"],var.image_re,"$2")}"
+    torcx_skip_setup         = "${var.tectonic_vanilla_k8s == 1 ? "true" : "false" }"
```
Comment (s-urbaniak) [inline modules/ignition/assets.tf]: `${var.tectonic_vanilla_k8s ? "true" : "false" }`

File: modules/ignition/variables.tf [hunk 3]
```diff
@@ -48,3 +48,23 @@ variable "cloud_provider_config" {
...
+}
+
+variable "torcx_store_url" {
+  type        = "string"
+  description = "(optional) URL template for torcx store. Leave empty to use the default CoreOS endpoint."
+  default     = ""
+}
+
+variable "tectonic_vanilla_k8s" {
+  default = false
```
Comment (s-urbaniak) [inline modules/ignition/variables.tf]: let's remove `default` here to enforce variable declaration.

File: config.tf [hunk 4]
```diff
@@ -82,6 +82,7 @@ variable "tectonic_container_images" {
     tectonic_etcd_operator       = "quay.io/coreos/tectonic-etcd-operator:v0.0.2"
     tectonic_prometheus_operator = "quay.io/coreos/tectonic-prometheus-operator:v1.6.0"
     tectonic_cluo_operator       = "quay.io/coreos/tectonic-cluo-operator:v0.2.0"
+    tectonic_torcx               = "quay.io/coreos/tectonic-torcx:installer-latest"
```
Comment (s-urbaniak) [inline config.tf]: `installer-latest`: do we want this as a final value?

Comment (lucab) [reply config.tf]: Yes, this is a mutable tag dedicated to the installer. We need this here for forward compatibility until some higher level component takes care of the kubernetes->docker mapping.
@robszumski has a tracker item for this somewhere.

**Comment by lucab** [comment]
@s-urbaniak thanks for the review, nits addressed, CI ongoing.

**Comment by Quentin-M** [comment]
All tests are red. Despite this being a release+feature freeze blocker, this cannot be merged as is.

**Comment by aaronlevy** [comment]
cc @yifan-gu

**Comment by lucab** [comment]
@Quentin-M sorry, it was fine before and I mistakenly broke it when addressing review comments. Amended.

**Comment by Quentin-M** [comment]
Thank you so much!

File: Documentation/dev/node-bootstrap-flow.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+## Systemd units
+
+The following systemd units are deployed to a node by tectonic-installer and take part in the bootstrapping process:
+
+* `k8s-node-bootstrap.service` ensures node and assets freshness. It is automatically started on boot, can crash-loop, and it runs only during bootstrap
+* `kubelet.service` is the main kubelet deamon. It is automatically started on boot, it is crash-looping until `kubelet.env` is populated, and it runs on each boot
+
+Additionally, only on one of the master node the following kubernetes bootstrapping happens:
+
+* `bootkube.service` deploys the initial bootstrapping control-plane. It is started only after ``kubelet.service` _is started_. It is a oneshot unit and cannot crash, and it runs only during bootstrap
```
Comment (diegs) [inline Documentation/dev/node-bootstrap-flow.md]: nit: double back-tick on `kubelet.service`

File: Documentation/dev/node-bootstrap-flow.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+1. `tectonic.service` is completed with success
+
+## Systemd units
+
+The following systemd units are deployed to a node by tectonic-installer and take part in the bootstrapping process:
+
+* `k8s-node-bootstrap.service` ensures node and assets freshness. It is automatically started on boot, can crash-loop, and it runs only during bootstrap
+* `kubelet.service` is the main kubelet deamon. It is automatically started on boot, it is crash-looping until `kubelet.env` is populated, and it runs on each boot
+
+Additionally, only on one of the master node the following kubernetes bootstrapping happens:
```
Comment (diegs) [inline Documentation/dev/node-bootstrap-flow.md]: nit: one of the master nodes

**Comment by Quentin-M** [comment]
Pushed to fix conflict in config.tf.. but looks like there are some new comments anyways. We can wait until tomorrow then.

**Comment by diegs** [comment]
@Quentin-M mine are just nits, they aren't release blocking and can be fixed in a follow-up PR.

File: modules/ignition/resources/services/k8s-node-bootstrap.service [hunk 7]
```diff
@@ -0,0 +1,36 @@
+[Unit]
+Description=Determine the Kubelet Image Version
```
Comment (yifan-gu) [inline modules/ignition/resources/services/k8s-node-bootstrap.service]: It does more than that now right? @lucab

**Comment by lucab** [comment]
@diegs thanks for proofreading!

@yifan-gu yes, it does docker setup (via torcx) and kubelet.env setup (similar to previous kube-version). I renamed the service but forgot to update the description.

@Quentin-M thanks for the push, it hit some flakes anyway. I rebased once more for the remaining nits, no flakes this time.

**Pull request merged**